components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java

```diff
@@ -759,12 +760,39 @@ private UserRecoveryData validateSelfRegistrationCode(String code, String verifi
         HashMap<String, String> userClaims = getClaimsListToUpdate(user, verifiedChannelType,
                 externallyVerifiedClaim, recoveryData.getRecoveryScenario().toString());
 
+        boolean supportMultipleEmailsAndMobileNumbers = Utils.isMultiEmailsAndMobileNumbersPerUserEnabled();
+        String multiAttributeSeparator = FrameworkUtils.getMultiAttributeSeparator();
+
         if (RecoverySteps.VERIFY_EMAIL.equals(recoveryData.getRecoveryStep())) {
             String pendingEmailClaimValue = recoveryData.getRemainingSetIds();
             if (StringUtils.isNotBlank(pendingEmailClaimValue)) {
                 eventProperties.put(IdentityEventConstants.EventProperty.VERIFIED_EMAIL, pendingEmailClaimValue);
                 userClaims.put(IdentityRecoveryConstants.EMAIL_ADDRESS_PENDING_VALUE_CLAIM, StringUtils.EMPTY);
-                userClaims.put(IdentityRecoveryConstants.EMAIL_ADDRESS_CLAIM, pendingEmailClaimValue); //todo??
+                if (supportMultipleEmailsAndMobileNumbers) {
+                    try {
+                        List<String> verifiedEmails = Utils.getMultiValuedClaim(userStoreManager, user,
+                                IdentityRecoveryConstants.VERIFIED_EMAIL_ADDRESSES_CLAIM);
+                        verifiedEmails.add(pendingEmailClaimValue);
+                        userClaims.put(IdentityRecoveryConstants.VERIFIED_EMAIL_ADDRESSES_CLAIM, StringUtils.join(
+                                verifiedEmails, multiAttributeSeparator));
+
+                        List<String> allEmails = Utils.getMultiValuedClaim(userStoreManager, user,
+                                IdentityRecoveryConstants.EMAIL_ADDRESSES_CLAIM);
+                        if (!allEmails.contains(pendingEmailClaimValue)) {
```

Comment:
Can allEmails be null?

Reply:
No `Utils.getMultiValuedClaim` always return ArrayList.